### PR TITLE
SNOW-1818018: add functions.size, collect_list

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1818018, SNOW-1825962

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [x] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://docs.google.com/document/d/162d_i4zZ2AfcGRXojj0jByt8EUq-DrSHPPnTa4QvwbA/edit#bookmark=id.e82u4nekq80k)

3. Please describe how your code solves the related issue.

   Add `functions.size` to get size of array, object or map column.
   Add `functions.collect_list` which is an alias to `functions.array_agg`.

### CODE REVIEW

File: src/snowflake/snowpark/functions.py [hunk 1]
```diff
@@ -5853,6 +5853,33 @@ def array_unique_agg(col: ColumnOrName) -> Column:
...
+        <BLANKLINE>
+    """
+    c = _to_col_if_str(col, "size")
+    v = to_variant(c)
+    return (
+        when(is_array(v), array_size(v))
+        .when(is_object(v), array_size(object_keys(v)))
+        .otherwise(lit(None))
+        .alias(f"SIZE({c.get_name()})")
+    )
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/functions.py:5880]: I'll check with SAS and make a request if necessary. This once seems useful

File: src/snowflake/snowpark/functions.py [hunk 2]
```diff
@@ -6722,6 +6722,42 @@ def array_unique_agg(col: ColumnOrName, _emit_ast: bool = True) -> Column:
...
+        ----------------------------------------------------------
+        |"SIZE(""COL1"")"  |"SIZE(""COL2"")"  |"SIZE(""COL3"")"  |
+        ----------------------------------------------------------
+        |3                 |2                 |NULL              |
+        ----------------------------------------------------------
+        <BLANKLINE>
+    """
+    c = _to_col_if_str(col, "size")
+    v = to_variant(c)
+    return (
```
Comment (sfc-gh-lspiegelberg) [inline src/snowflake/snowpark/functions.py:6745]: same comment as in https://github.com/snowflakedb/snowpark-python/pull/2682 wrt to JIRA, and also using `_emit_ast=False` for internal logic calls.

File: tests/integ/test_function.py [hunk 3]
```diff
@@ -1742,6 +1743,10 @@ def test_array_negative(session):
         df.select(array_size([1])).collect()
     assert "'ARRAY_SIZE' expected Column or str, got: <class 'list'>" in str(ex_info)
 
+    with pytest.raises(TypeError) as ex_info:
```
Comment (sfc-gh-lspiegelberg) [inline tests/integ/test_function.py:1746]: use instead `pytest.raises(TypeError, matches='...')`. You do not need the assert then, much easier to read and maintain.

File: tests/integ/test_function.py [hunk 4]
```diff
@@ -1742,6 +1743,10 @@ def test_array_negative(session):
...
+        df.select(size([1])).collect()
+    assert "'SIZE' expected Column or str, got: <class 'list'>" in str(ex_info)
+
     with pytest.raises(TypeError) as ex_info:
         df.select(array_slice([1], "col1", "col2")).collect()
```
Comment (sfc-gh-lspiegelberg) [inline tests/integ/test_function.py:1751]: nit: Where are the tests for the new function and all of its code paths?

Comment (sfc-gh-aalam) [reply tests/integ/test_function.py:1751]: they are added as doctests

File: tests/integ/test_function.py (hunk 3 above)
Comment (sfc-gh-aalam) [reply tests/integ/test_function.py:1746]: I can update them.

**Review by sfc-gh-lspiegelberg — APPROVED** [review]
LGTM!

**Pull request merged**